File: src/valicore/campaign/runner.py
from __future__ import annotations

import time
from datetime import datetime, timezone
from typing import Any, Callable

from valicore.campaign.model import TestCampaign, ComparisonOp
from valicore.core import RustSignalProcessor
from valicore.instruments.base import SCPIInstrument


def _eq(value: float, target: float, tol: float | None) -> bool:
    return abs(value - target) <= (tol or 1e-9)


def _ne(value: float, target: float, tol: float | None) -> bool:
    return abs(value - target) > (tol or 1e-9)


def _lt(value: float, target: float, _tol: float | None) -> bool:
    return value < target


def _le(value: float, target: float, _tol: float | None) -> bool:
    return value <= target


def _gt(value: float, target: float, _tol: float | None) -> bool:
    return value > target


def _ge(value: float, target: float, _tol: float | None) -> bool:
    return value >= target


def _within(value: float, target: float, tol: float | None) -> bool:
    return abs(value - target) <= (tol or 0.0)


def _outside(value: float, target: float, tol: float | None) -> bool:
    return abs(value - target) > (tol or 0.0)


_CHECK_LIMIT_FUNCS: dict[ComparisonOp, Callable[[float, float, float | None], bool]] = {
    ComparisonOp.eq: _eq,
    ComparisonOp.ne: _ne,
    ComparisonOp.lt: _lt,
    ComparisonOp.le: _le,
    ComparisonOp.gt: _gt,
    ComparisonOp.ge: _ge,
    ComparisonOp.within: _within,
    ComparisonOp.outside: _outside,
}


class CampaignRunner:
    def __init__(self, campaign: TestCampaign, instrument_overrides: dict[str, str] | None = None):
        self.campaign = campaign
        self.results: dict[str, Any] = {
            "title": campaign.title,
            "version": campaign.version,
            "timestamp": datetime.now(timezone.utc).isoformat(),
            "groups": {},
        }
        self._instruments: dict[str, SCPIInstrument] = {}
        self._instrument_overrides = instrument_overrides or {}
        self._signal = RustSignalProcessor()

    def run(self) -> dict[str, Any]:
        groups = self.results["groups"]
        for group_name, group in self.campaign.groups.items():
            groups[group_name] = self._run_group(group_name, group)
        return self.results

    def _run_group(self, group_name: str, group: Any) -> dict[str, Any]:
        group_result: dict[str, Any] = {
            "name": group.name,
            "description": group.description,
            "status": "skipped",
            "steps": [],
        }

        if group.setup:
            self._execute_setup(group.setup)

        steps = []
        for step in group.steps:
            step_result = self._run_step(step)
            steps.append(step_result)
            if step_result["status"] == "failed":
                break

        group_result["steps"] = steps
        all_passed = True
        for s in steps:
            if s["status"] != "passed":
                all_passed = False
                break
        group_result["status"] = "passed" if all_passed else "failed"

        if group.teardown:
            self._execute_teardown(group.teardown)

        return group_result

    def _run_step(self, step: Any) -> dict[str, Any]:
        step_result: dict[str, Any] = {
            "name": step.name,
            "description": step.description,
            "status": "passed",
            "measurements": [],
            "error": None,
        }

        try:
            instr = self._get_instrument(step.instrument)

            if step.repeat > 0:
                delay_s = step.delay_ms / 1000.0
                for i in range(step.repeat):
                    if i > 0 and delay_s:
                        time.sleep(delay_s)

                    raw_data = None
                    if step.command:
                        instr.write(step.command)
                        raw_data = instr.query(step.command)

                    measure_funcs = _CHECK_LIMIT_FUNCS
                    for meas in step.measurements:
                        value = self._acquire_measurement(instr, meas, raw_data)
                        if step.post_processing:
                            value = self._post_process(value, step.post_processing)
                        verdict = "passed"
                        if meas.limits:
                            for l in meas.limits:
                                check = measure_funcs.get(ComparisonOp(l.op))
                                if check is not None and not check(value, l.value, l.tolerance):
                                    verdict = "failed"
                                    break
                        step_result["measurements"].append(
                            {
                                "name": meas.name,
                                "value": value,
                                "limits": [l.model_dump() for l in meas.limits],
                                "verdict": verdict,
                            }
                        )
                        if verdict != "passed":
                            step_result["status"] = "failed"

        except Exception as e:
            step_result["status"] = "failed"
            step_result["error"] = str(e)

        return step_result

    def _get_instrument(self, name: str) -> SCPIInstrument:
        instruments = self._instruments
        cached = instruments.get(name)
        if cached is not None:
            return cached

        config = self.campaign.instruments.get(name)
        if not config:
            raise ValueError(f"instrument '{name}' not defined in campaign")

        resource = self._instrument_overrides.get(name, config.resource)
        instr = SCPIInstrument(resource, timeout=config.timeout)
        instruments[name] = instr
        return instr

    def _acquire_measurement(self, instr: SCPIInstrument, meas: Any, raw_data: Any) -> float:
        if meas.command:
            response = instr.query(meas.command)
            return float(response.strip())
        if raw_data is not None:
            parts = raw_data.strip().split(",")
            return float(parts[0])
        return 0.0

    def _post_process(self, value: Any, pp: Any) -> Any:
        if isinstance(value, list):
            samples = value
        else:
            samples = [float(value)]

        if pp.window:
            samples = self._signal.window(samples, pp.window)
        if pp.filter_type:
            samples = self._signal.filter(
                samples,
                filter_type=pp.filter_type,
                cutoff=pp.filter_cutoff or 0.5,
                order=pp.filter_order or 2,
            )
        if samples:
            return samples[0]
        return value

    def _check_limit(self, value: float, limit: Any) -> bool:
        check = _CHECK_LIMIT_FUNCS.get(ComparisonOp(limit.op))
        if check is None:
            return False
        return check(value, limit.value, limit.tolerance)

    def close(self) -> None:
        for instr in self._instruments.values():
            try:
                instr.close()
            except Exception:
                pass

File: src/engine/runner.rs
use std::collections::HashMap;
use std::sync::Arc;

use serde_json::{json, Value};
use tokio::sync::Mutex;

use super::campaign::{Limit, TestCampaign, TestStep};
use super::instrument::{create_instrument, SCPIInstrument};

fn check_limit(value: f64, limit: &Limit) -> bool {
    let tol = limit.tolerance.unwrap_or(1e-9);
    match limit.op.as_str() {
        "eq" => (value - limit.value).abs() <= tol,
        "ne" => (value - limit.value).abs() > tol,
        "lt" => value < limit.value,
        "le" => value <= limit.value,
        "gt" => value > limit.value,
        "ge" => value >= limit.value,
        "within" => (value - limit.value).abs() <= tol,
        "outside" => (value - limit.value).abs() > tol,
        _ => true,
    }
}

fn evaluate_limits(value: f64, limits: &[Limit]) -> &'static str {
    if limits.iter().all(|l| check_limit(value, l)) {
        "passed"
    } else {
        "failed"
    }
}

fn parse_measurement(response: &str) -> f64 {
    if let Ok(v) = response.trim().parse::<f64>() {
        return v;
    }
    response
        .split(',')
        .next()
        .and_then(|s| s.trim().parse::<f64>().ok())
        .unwrap_or(0.0)
}

type InstrumentMap = Arc<Mutex<HashMap<String, Box<dyn SCPIInstrument>>>>;

async fn ensure_instrument(
    step_instr_name: &str,
    campaign: &TestCampaign,
    pool: &InstrumentMap,
) -> Result<(), String> {
    let mut map = pool.lock().await;
    if map.contains_key(step_instr_name) {
        return Ok(());
    }
    let config = campaign
        .instruments
        .get(step_instr_name)
        .ok_or_else(|| format!("instrument '{}' not defined in campaign", step_instr_name))?;

    let timeout = config.timeout.unwrap_or(5000);
    let mut instr = create_instrument(&config.kind, &config.resource, timeout);
    instr
        .connect()
        .await
        .map_err(|e| format!("connect {}: {}", step_instr_name, e))?;
    map.insert(step_instr_name.to_string(), instr);
    Ok(())
}

async fn query_instrument(
    step_instr_name: &str,
    cmd: &str,
    pool: &InstrumentMap,
) -> Result<String, String> {
    let mut map = pool.lock().await;
    let instr = map
        .get_mut(step_instr_name)
        .ok_or_else(|| format!("instrument '{}' not connected", step_instr_name))?;
    instr.query(cmd).await.map_err(|e| format!("query error: {}", e))
}

async fn run_step(
    step: &TestStep,
    campaign: &TestCampaign,
    pool: &InstrumentMap,
) -> Value {
    let mut step_result = json!({
        "name": step.name,
        "description": step.description,
        "status": "passed",
        "measurements": [],
        "error": null,
    });

    if let Err(e) = ensure_instrument(&step.instrument, campaign, pool).await {
        if let Some(obj) = step_result.as_object_mut() {
            obj.insert("status".into(), json!("failed"));
            obj.insert("error".into(), json!(e));
        }
        return step_result;
    }

    for _ in 0..step.repeat {
        if step.delay_ms > 0 {
            tokio::time::sleep(tokio::time::Duration::from_millis(step.delay_ms)).await;
        }

        let cmd = step.command.as_deref().unwrap_or("*IDN?");
        let response = match query_instrument(&step.instrument, cmd, pool).await {
            Ok(r) => r,
            Err(e) => {
                if let Some(obj) = step_result.as_object_mut() {
                    obj.insert("status".into(), json!("failed"));
                    obj.insert("error".into(), json!(e));
                }
                return step_result;
            }
        };

        for meas in &step.measurements {
            let value = parse_measurement(&response);
            let verdict = meas
                .limits
                .as_ref()
                .map_or("passed", |l| evaluate_limits(value, l));

            let meas_result = json!({
                "name": meas.name,
                "value": value,
                "limits": meas.limits.as_ref().map(|l| {
                    json!(l.iter().map(|lim| {
                        json!({
                            "op": lim.op,
                            "value": lim.value,
                            "tolerance": lim.tolerance,
                        })
                    }).collect::<Vec<_>>())
                }),
                "verdict": verdict,
            });

            if let Some(arr) = step_result.get_mut("measurements").and_then(|v| v.as_array_mut()) {
                arr.push(meas_result);
            }

            if verdict == "failed" {
                if let Some(obj) = step_result.as_object_mut() {
                    obj.insert("status".into(), json!("failed"));
                }
            }
        }
    }

    step_result
}

pub async fn run_campaign(campaign: &TestCampaign) -> Result<Value, String> {
    let pool: InstrumentMap = Arc::new(Mutex::new(HashMap::new()));
    let mut groups = serde_json::Map::new();

    for (group_name, group) in &campaign.groups {
        let mut steps = Vec::new();
        for step in &group.steps {
            let result = run_step(step, campaign, &pool).await;
            steps.push(result);
        }

        let status = if steps.iter().any(|s| s["status"] == "failed") {
            "failed"
        } else {
            "passed"
        };

        groups.insert(
            group_name.clone(),
            json!({
                "name": group.name,
                "description": group.description,
                "status": status,
                "steps": steps,
            }),
        );
    }

    let now = std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .unwrap_or_default()
        .as_secs();

    let results = json!({
        "title": campaign.title,
        "version": campaign.version.as_deref().unwrap_or("1.0"),
        "timestamp": now.to_string(),
        "groups": groups,
    });

    Ok(results)
}

pub fn format_summary(results: &Value) -> String {
    let total: usize = results
        .pointer("/groups")
        .and_then(|v| v.as_object())
        .map(|g| {
            g.values()
                .filter_map(|v| v.get("steps").and_then(|s| s.as_array()))
                .map(|s| s.len())
                .sum()
        })
        .unwrap_or(0);
    let failed: usize = results
        .pointer("/groups")
        .and_then(|v| v.as_object())
        .map(|g| {
            g.values()
                .filter_map(|v| v.get("steps").and_then(|s| s.as_array()))
                .flatten()
                .filter(|s| s.get("status").and_then(|v| v.as_str()) == Some("failed"))
                .count()
        })
        .unwrap_or(0);
    format!("{}/{} passed, {} failed", total - failed, total, failed)
}

use tokio_stream::wrappers::UnboundedReceiverStream;
use tokio_stream::Stream;

pub fn run_campaign_stream(campaign: TestCampaign) -> impl Stream<Item = (String, Value)> {
    let (tx, rx) = tokio::sync::mpsc::unbounded_channel();

    tokio::spawn(async move {
        let pool: InstrumentMap = Arc::new(Mutex::new(HashMap::new()));

        for (group_name, group) in &campaign.groups {
            let mut steps = Vec::new();
            for step in &group.steps {
                let result = run_step(step, &campaign, &pool).await;
                steps.push(result);
            }

            let status = if steps.iter().any(|s| s["status"] == "failed") {
                "failed"
            } else {
                "passed"
            };

            let group_result = json!({
                "name": group.name,
                "description": group.description,
                "status": status,
                "steps": steps,
            });

            let _ = tx.send((group_name.clone(), group_result));
        }
    });

    UnboundedReceiverStream::new(rx)
}
